### PR TITLE
feat: add multiple save for redis registry

### PR DESCRIPTION
**Pull request by s0nicboOm**
1) Option to multisave in redis registry
2) support caching for version calls
3) In udfs/ load models using version call instead of latest calls

### CODE REVIEW

File: numalogic/config/_config.py [hunk 1]
```diff
@@ -83,7 +83,9 @@ class TrainerConf:
 class NumalogicConf:
     """Top level config schema for numalogic."""
 
-    model: ModelInfo = field(default_factory=ModelInfo)
+    model: ModelInfo = field(
+        default_factory=lambda: ModelInfo(name="VanillaAE", conf={"n_features": 2, "seq_len": 10})
+    )
```
Comment (ab93) [inline numalogic/config/_config.py:88]: Any reason why are we doing this? We should not have a default for the number of features arg

File: numalogic/connectors/druid.py [hunk 2]
```diff
@@ -48,7 +48,7 @@ def fetch_data(
 
         end_dt = datetime.now(pytz.utc)
         start_dt = end_dt - timedelta(hours=hours)
-        intervals = f"{start_dt.isoformat()}/{end_dt.isoformat()}"
+        intervals = [f"{start_dt.isoformat()}/{end_dt.isoformat()}"]
```
Comment (ab93) [inline numalogic/connectors/druid.py:51]: This is with regards to Chichu's fork correct?

File: numalogic/udfs/entities.py [hunk 3]
```diff
@@ -81,6 +81,9 @@ def get_data(self, original=False) -> npt.NDArray[float]:
     def get_metadata(self, key: str) -> dict[str, Any]:
         return copy(self.metadata[key])
 
+    def add_metadata(self, data: dict[str, Any]):
+        self.metadata.update(data)
```
Comment (ab93) [inline numalogic/udfs/entities.py:85]: Let's add a test case for this to improve coverage?

File: numalogic/udfs/__main__.py [hunk 4]
```diff
@@ -12,8 +12,13 @@
 CONF_FILE_PATH = os.getenv(
     "CONF_PATH", default=os.path.join(BASE_CONF_DIR, "default-configs", "config.yaml")
 )
+pipeline_conf = load_pipeline_conf(CONF_FILE_PATH)
+logging.info("Pipeline config: %s", pipeline_conf)
 
+redis_client = get_redis_client_from_conf(pipeline_conf.redis_conf)
 
+for key in redis_client.scan_iter("*"):
+    redis_client.delete(key)
```
Comment (ab93) [inline numalogic/udfs/__main__.py:21]: Why are we deleting all the keys here?

**Review by ab93 — CHANGES_REQUESTED** [review]
Some comments

File: numalogic/udfs/tools.py [hunk 5]
```diff
@@ -107,6 +121,22 @@ def _load_model(
...
-        return artifact
+        _LOGGER.info(
+            "%s - Loaded Model. Source: %s , version: %s, Keys: %s, %s",
+            payload.uuid,
+            artifact.extras.get("source"),
+            artifact.extras.get("version"),
+            skeys,
+            dkeys,
+        )
+        if artifact.metadata and "artifact_versions" in artifact.metadata:
```
Comment (nkoppisetty) [inline numalogic/udfs/tools.py:134]: shouldn't this be updated only when there is not artifact_versions in the payload metadata?

File: tests/udfs/test_inference.py [hunk 6]
```diff
@@ -83,14 +84,15 @@ class TestInferenceUDF(unittest.TestCase):
...
+        print(self.udf.get_conf("conf1"))
 
     @patch.object(
         RedisRegistry,
         "load",
         Mock(
             return_value=ArtifactData(
                 artifact=VanillaAE(seq_len=12, n_features=2),
-                extras=dict(version="1", timestamp=time.time(), source="registry"),
+                extras=dict(version="0", timestamp=time.time(), source="registry"),
```
Comment (nkoppisetty) [inline tests/udfs/test_inference.py:95]: Add test cases to assert we are loading correct versions for all the 3 models, also try to add edge cases in this scenario

**Review by nkoppisetty — CHANGES_REQUESTED** [review]
couple of comments, and also you need to rebase.

File: numalogic/udfs/__main__.py (hunk 4 above)
Comment (s0nicboOm) [reply numalogic/udfs/__main__.py:21]: removed..
This was added for testing

File: numalogic/config/_config.py (hunk 1 above)
Comment (s0nicboOm) [reply numalogic/config/_config.py:88]: I was keeping everything to have a default. Wanted to get your thoughts on this too.
Can revert it back if needed

File: numalogic/udfs/tools.py (hunk 5 above)
Comment (s0nicboOm) [reply numalogic/udfs/tools.py:134]: Even though the artifact metadata is same foe all 3 set of models but yes I can add another if to remove redundant update of artifact_version.
Good catch!

File: tests/udfs/test_inference.py (hunk 6 above)
Comment (s0nicboOm) [reply tests/udfs/test_inference.py:95]: I think testing for all 3 models is out of scope of unit testing.
Let me know what are your thoughts.

File: numalogic/registry/redis_registry.py [hunk 7]
```diff
@@ -155,30 +159,37 @@ def __load_latest_artifact(self, key: str) -> tuple[ArtifactData, bool]:
...
-        return self.__get_artifact_data(
-            model_key=model_key,
+    def __load_version_artifact(self, version: str, key: str) -> tuple[ArtifactData, bool]:
+        version_key = self.__construct_version_key(key, version)
+        cached_artifact = self._load_from_cache(version_key)
+        if cached_artifact:
+            _LOGGER.debug("Found cached version artifact for key: %s", version_key)
+            return cached_artifact, True
+        if not self.client.exists(version_key):
+            raise ModelKeyNotFound("Could not find model key with key: %s" % version_key)
```
Comment (ab93) [inline numalogic/registry/redis_registry.py:183]: You can use f-strings here. We use C style str formatting only when logging

File: numalogic/registry/redis_registry.py [hunk 8]
```diff
@@ -230,24 +243,33 @@ def load(
...
+                    )
+                    self._save_in_cache(self.__construct_version_key(key, version), artifact_data)
             return artifact_data
 
     def save(
         self,
         skeys: KEYS,
         dkeys: KEYS,
         artifact: artifact_t,
+        pipe=None,
```
Comment (ab93) [inline numalogic/registry/redis_registry.py:272]:
```suggestion
        _pipe=None,
```
If this will never be called by the end user

Comment (ab93) [reply numalogic/registry/redis_registry.py:272]: Also add to docstrings

File: numalogic/tools/types.py [hunk 9]
```diff
@@ -40,6 +38,13 @@
 KEYS = TypeVar("KEYS", bound=Sequence[str], covariant=True)
 
 
+class ArtifactTuple(NamedTuple):
```
Comment (ab93) [inline numalogic/tools/types.py:41]:
```suggestion
class KeyedArtifact(NamedTuple):
```

File: numalogic/registry/redis_registry.py [hunk 10]
```diff
@@ -338,3 +365,48 @@ def is_artifact_stale(artifact_data: ArtifactData, freq_hr: int) -> bool:
...
+                for key, value in dict_artifacts.items():
+                    dict_model_ver[":".join(value.dkeys)] = self.save(
+                        skeys=skeys,
+                        dkeys=value.dkeys,
+                        artifact=value.artifact,
+                        pipe=pipe,
+                        **metadata,
+                    )
+
+                    if len(dict_artifacts) == len(dict_model_ver):
```
Comment (ab93) [inline numalogic/registry/redis_registry.py:398]: `else` didn't work here?

File: numalogic/registry/redis_registry.py [hunk 11]
```diff
@@ -230,24 +243,33 @@ def load(
...
-        is_cached = False
         try:
             if latest:
                 artifact_data, is_cached = self.__load_latest_artifact(key)
             else:
-                artifact_data = self.__load_version_artifact(version, key)
+                artifact_data, is_cached = self.__load_version_artifact(version, key)
         except RedisError as err:
             raise RedisRegistryError(f"{err.__class__.__name__} raised") from err
         else:
```
Comment (ab93) [inline numalogic/registry/redis_registry.py:253]: You can remove this else I think

File: numalogic/registry/redis_registry.py [hunk 12]
```diff
@@ -230,24 +243,33 @@ def load(
...
-            if (not is_cached) and latest:
-                self._save_in_cache(key, artifact_data)
+            if not is_cached:
+                if latest:
+                    _LOGGER.info("Saving %s, in cache as %s", self.__construct_latest_key(key), key)
```
Comment (ab93) [inline numalogic/registry/redis_registry.py:256]: Should be debug log right?

File: numalogic/udfs/tools.py [hunk 13]
```diff
@@ -107,6 +121,26 @@ def _load_model(
...
-        return artifact
+        _LOGGER.info(
+            "%s - Loaded Model. Source: %s , version: %s, Keys: %s, %s",
+            payload.uuid,
+            artifact.extras.get("source"),
+            artifact.extras.get("version"),
+            skeys,
+            dkeys,
+        )
+        if (
```
Comment (ab93) [inline numalogic/udfs/tools.py:134]: add a TODO here to not mutate the payload here

**Review by ab93 — COMMENTED** [review]
Few comments

File: numalogic/registry/redis_registry.py (hunk 8 above)
Comment (s0nicboOm) [reply numalogic/registry/redis_registry.py:272]: Yes. Let me add that

File: numalogic/tools/types.py (hunk 9 above)
Comment (s0nicboOm) [reply numalogic/tools/types.py:41]: Done

File: numalogic/registry/redis_registry.py (hunk 10 above)
Comment (s0nicboOm) [reply numalogic/registry/redis_registry.py:398]: We want to execute this statement in the last iteration of loop

File: numalogic/udfs/tools.py (hunk 13 above)
Comment (s0nicboOm) [reply numalogic/udfs/tools.py:134]: Sure

File: numalogic/registry/redis_registry.py (hunk 7 above)
Comment (s0nicboOm) [reply numalogic/registry/redis_registry.py]: good point. Fixed!

**Review by ab93 — APPROVED** [review]
Please rebase before merging

**Review by nkoppisetty — APPROVED** [review]
LGTM
